crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -271,13 +272,26 @@ impl CheckpointExecutor {
     async fn schedule_checkpoint(
         &mut self,
         checkpoint: VerifiedCheckpoint,
-        finished: &mut CheckpointExecutionBuffer,
+        pending: &mut CheckpointExecutionBuffer,
     ) -> SuiResult {
-        // Since StateSync is guaranteed to produce checkpoint sync messages
-        // in monotonically increasing order, then a mismatch of epoch would
-        // mean that we failed to execute reconfig at the epoch boundary. This
-        // is an invalid state that should not happen under normal operation
-        assert_eq!(checkpoint.epoch(), self.authority_state.epoch());
+        // Mismatch between node epoch and checkpoit epoch only valid if we crashed after
+        // reconfig but before updating highest_executed_checkpoint watermark. In this case,
+        // we can easily recover by reconfiging now since there should be no scheduled checkpoints.
+        if checkpoint.epoch() != self.authority_state.epoch() {
+            assert!(
+                pending.is_empty() && self.highest_scheduled_seq_num.is_none(),
+                "Mismatch between node epoch and checkpoint epoch is only valid at startup"
+            );
+            let new_epoch_committee = self
+                .checkpoint_store
+                .get_highest_executed_checkpoint()?
+                .unwrap()
+                .next_epoch_committee()
+                .unwrap_or_else(
+                    || panic!("Expected last executed checkpoint to be last of epoch due to epoch mismatch, bit it was not")
+                ).to_vec();
+            reconfig(new_epoch_committee).await;
```

Comment:
Since this should only happen due to crash recovery, what if we lifted this to the very start of the `run` function?

Reply:
Because we don't know until we get the first checkpoint